src/base/bittorrent/tracker.cpp

```diff
@@ -107,7 +107,7 @@ namespace
         case QAbstractSocket::IPv6Protocol: {
                 const Q_IPV6ADDR ipv6 = addr.toIPv6Address();
                 QByteArray ret;
-                for (int i = (sizeof(ipv6.c) - 1); i >= 0; --i)
+                for (unsigned long i = 0; i < sizeof(ipv6.c); ++i)
```

Comment:
Use signed integer for looping index. You should revert to `int`.

Reply:
That results in `warning: comparison between signed and unsigned integer expressions [-Wsign-compare]`, since `sizeof()` returns an `unsigned long`.

Reply:
Actually I should `std::size_t` which is more correct.

Reply:
Please don't complicate the issue... also IIRC that compile flag is not defined in qbt. It is too verbose to be really useful.

Reply:
I compile qBittorrent with default options and I get that warning. The commands I use are simply:

```bash
./configure --enable-debug
make -j2
```

Reply:
A more compelling answer not to use unsigned numbers in this case: https://www.reddit.com/r/cpp/comments/7y0o6r/is_it_a_good_practice_to_use_unsigned_values/ductv0f/

Reply:
@Chocobo1 
In the linked video your link leads to, at approximately 13:05 (https://www.youtube.com/watch?v=Puio5dly9N8&feature=youtu.be&t=785), Stroustrup states "never make signed an unsigned". This is exactly what would happen, and it's also what the warning is about:

>-Wsign-compare
            Warn when a comparison between signed and unsigned values could produce an incorrect result when the signed value is converted to unsigned.

This is also corroborated by https://en.cppreference.com/w/cpp/types/size_t

> When indexing C++ containers, such as std::string, std::vector, etc, the appropriate type is the member typedef size_type provided by such containers. It is usually defined as a synonym for std::size_t.

I understand the arguments against using unsigned values, but this would be one of the cases where it would be correct IMO. We are iterating forward over a container _and_ not doing any math with the index variable.

Reply:
>Stroustrup states "never make signed an unsigned". This is exactly what would happen

I think you didn't really catch what he really means. Really, it is all summarized up in that reddit link. Unsigned number should only be used in the mentioned cases because there are too many ways to loose control of it.

>but this would be one of the cases where it would be correct IMO

If that is one valid case, then in the reddit link would mention it. But it didn't.

Another way is to narrow the sizeof() to int. But I also think it is too verbose to write out such a cast just for the sake of a compiler warning (I can't really oppose if you insist this way).

Reply:
>In the linked video your link leads to, at approximately 13:05 (timestamped link: https://www.youtube.com/watch?v=Puio5dly9N8&feature=youtu.be&t=785), 

Before you start finding evidences to back you up, at least watch all of the time points in the reddit link, especially the one near 1:02:50.

Reply:
>I compile qBittorrent with default options and I get that warning. 

FYI, I tested on my machine and have no such warning and neither does our CI.

Reply:
> FYI, I tested on my machine and have no such warning and neither does our CI.

I am using Ubuntu 18.04 with latest updates and HWE:
```
$ uname -a
Linux xubuntu-vm 5.3.0-28-generic #30~18.04.1-Ubuntu SMP Fri Jan 17 06:14:09 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
$ g++ --version
g++ (Ubuntu 8.3.0-6ubuntu1~18.04.1) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
Earlier I forgot to mention I am actually also using boost compiled from source (1.72.0), but I don't think that matters:

Still, for reference, boost was compiled with:

```
./bootstrap.sh --prefix=/opt/boost  --with-python=/usr/bin/python3
sudo ./b2 install -a -q -d+2 --prefix=/opt/boost --with=all -j2
```

As for qBIttorrent, here are the actual commands my system is using to compile `tracker.cpp:`

```
$ ./configure --enable-debug --with-boost=/opt/boost
<output omitted>
```
```
$ make -n | grep tracker.cpp
echo compiling base/bittorrent/tracker.cpp && g++ -c -pipe -g -O2 -fexceptions -std=c++14 -g -O2 -Wall -Wextra -Wpedantic -Wformat-security -O2 -D_REENTRANT -Wall -W -fPIC -DTORRENT_LINKING_SHARED -DBOOST_ASIO_ENABLE_CANCELIO -DTORRENT_USE_ICONV -DTORRENT_USE_OPENSSL -DTORRENT_USE_LIBCRYPTO -DBOOST_SYSTEM_DYN_LINK -DBOOST_ALL_NO_LIB -DSTACKTRACE -DQT_NO_DEBUG_OUTPUT -DQBT_VERSION_MAJOR=4 -DQBT_VERSION_MINOR=3 -DQBT_VERSION_BUGFIX=0 -DQBT_VERSION_BUILD=0 -DQBT_VERSION=\"v4.3.0alpha1\" -DQBT_VERSION_2=\"4.3.0alpha1\" -DQT_DEPRECATED_WARNINGS -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DQT_STRICT_ITERATORS -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_DBUS_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_CORE_LIB -I. -I/opt/boost/include -I/opt/boost/include -I. -Iapp -Igui -Igui/powermanagement -Igui/properties -Igui/qtnotify -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtSvg -isystem /usr/include/x86_64-linux-gnu/qt5/QtWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtDBus -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I. -isystem /usr/include/libdrm -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o tracker.o base/bittorrent/tracker.cpp
echo moc base/bittorrent/tracker.h && /usr/lib/qt5/bin/moc -DTORRENT_LINKING_SHARED -DBOOST_ASIO_ENABLE_CANCELIO -DTORRENT_USE_ICONV -DTORRENT_USE_OPENSSL -DTORRENT_USE_LIBCRYPTO -DBOOST_SYSTEM_DYN_LINK -DBOOST_ALL_NO_LIB -DSTACKTRACE -DQT_NO_DEBUG_OUTPUT -DQBT_VERSION_MAJOR=4 -DQBT_VERSION_MINOR=3 -DQBT_VERSION_BUGFIX=0 -DQBT_VERSION_BUILD=0 -DQBT_VERSION=\"v4.3.0alpha1\" -DQBT_VERSION_2=\"4.3.0alpha1\" -DQT_DEPRECATED_WARNINGS -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DQT_STRICT_ITERATORS -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_DBUS_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_CORE_LIB --include ./moc_predefs.h -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -I/home/francisco/Documents/qBittorrent/src -I/opt/boost/include -I/opt/boost/include -I/home/francisco/Documents/qBittorrent/src -I/home/francisco/Documents/qBittorrent/src/app -I/home/francisco/Documents/qBittorrent/src/gui -I/home/francisco/Documents/qBittorrent/src/gui/powermanagement -I/home/francisco/Documents/qBittorrent/src/gui/properties -I/home/francisco/Documents/qBittorrent/src/gui/qtnotify -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtSvg -I/usr/include/x86_64-linux-gnu/qt5/QtWidgets -I/usr/include/x86_64-linux-gnu/qt5/QtGui -I/usr/include/x86_64-linux-gnu/qt5/QtDBus -I/usr/include/x86_64-linux-gnu/qt5/QtNetwork -I/usr/include/x86_64-linux-gnu/qt5/QtXml -I/usr/include/x86_64-linux-gnu/qt5/QtCore -I/usr/include/c++/7 -I/usr/include/x86_64-linux-gnu/c++/7 -I/usr/include/c++/7/backward -I/usr/lib/gcc/x86_64-linux-gnu/7/include -I/usr/lib/gcc/x86_64-linux-gnu/7/include-fixed -I/usr/include/x86_64-linux-gnu -I/usr/include base/bittorrent/tracker.h -o moc_tracker.cpp
echo compiling moc_tracker.cpp && g++ -c -pipe -g -O2 -fexceptions -std=c++14 -g -O2 -Wall -Wextra -Wpedantic -Wformat-security -O2 -D_REENTRANT -Wall -W -fPIC -DTORRENT_LINKING_SHARED -DBOOST_ASIO_ENABLE_CANCELIO -DTORRENT_USE_ICONV -DTORRENT_USE_OPENSSL -DTORRENT_USE_LIBCRYPTO -DBOOST_SYSTEM_DYN_LINK -DBOOST_ALL_NO_LIB -DSTACKTRACE -DQT_NO_DEBUG_OUTPUT -DQBT_VERSION_MAJOR=4 -DQBT_VERSION_MINOR=3 -DQBT_VERSION_BUGFIX=0 -DQBT_VERSION_BUILD=0 -DQBT_VERSION=\"v4.3.0alpha1\" -DQBT_VERSION_2=\"4.3.0alpha1\" -DQT_DEPRECATED_WARNINGS -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DQT_STRICT_ITERATORS -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_DBUS_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_CORE_LIB -I. -I/opt/boost/include -I/opt/boost/include -I. -Iapp -Igui -Igui/powermanagement -Igui/properties -Igui/qtnotify -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtSvg -isystem /usr/include/x86_64-linux-gnu/qt5/QtWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtDBus -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I. -isystem /usr/include/libdrm -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o moc_tracker.o moc_tracker.cpp
```

Notice the `-Wall` flag, which in C++ is the one that also enables `-Wsign-compare`. Don't know what you are using, but the CI is using Ubuntu 16.04. I checked the manpage for the gcc version it uses (it's something like 5.x) and it also says that `-Wall` activates `-Wsign-compare` for C++, so we should see a warning there. Unfortunately, I can't see the verbose compilation commands the CI is using, so I can't be sure of what it is doing.


> I think you didn't catch what he really means. Really, it is all summarized up in that reddit link. Unsigned number should only be used in the mentioned cases because there are too many ways to lose control of it.

From the context of the sentence, I interpret it as "never convert a singed to an unsigned". This is what happens in a comparison between a signed and an unsigned.

> Another way is to narrow the sizeof() to int. But I also think it is too verbose to write out such a cast just for the sake of a compiler warning (I can't really oppose if you insist this way).

I don't want to do it either.

> If that is one valid case, then in the reddit link would mention it. But it didn't.
> Before you start finding evidences to back you up, at least watch all of the time points in the reddit link, especially the one near 1:02:50.

I watched that part as well. The thing is, the state of the language and the standard library is such that for certain situations like this one there is not one right answer. The official "best practices" contradict themselves. We can't follow the advice of using signed integers for indexing without going against "don't do things that force (sometimes non-obvious) singed -> unsigned conversion". In this specific case, I think following the latter advice is more important.

By the way, for the sake of argument, the video also mentions another valid (albeit contrived) case where the advice of "don't use unsigned integers for indexing" is simply impossible to follow (44:37).

Still, if you insist that I change it back to `int`, let me know and I will do so.

Reply:
I just tested on Debian and I have this warning when using `int`

Reply:
>The thing is, the state of the language and the standard library is such that for certain situations like this one there is not one right answer. 

There is, you should follow the language at large and consider the library second. Just because the library don't want to break compatibility doesn't mean they are right.

>We can't follow the advice of using signed integers for indexing without going against "don't do things that force (sometimes non-obvious) singed -> unsigned conversion". 

You're such narrow minded, why don't you just narrow the unsigned to signed and everyone is happy?

>Still, if you insist that I change it back to int, let me know and I will do so.

Now I really played with it a bit (and saw the warning), I think the proper solution is to narrow the sizeof():
```c++
for (int i = 0; i < static_cast<int>(sizeof(ipv6.c)); ++i)
```
This is much safer than letting unsigned run amok in our codes.

Reply:
> You're such narrow minded, why don't you just narrow the unsigned to signed and everyone is happy?

Because I had the feeling that "there must be a better way". And in fact there was; the range-based for approach (thanks for that). You yourself expressed unwillingness to go the cast route in the first place...

Anyway, instead of:

```cpp
for (const quint8 i : ipv6.c)
    ret.append(i);
```

I can use the reference:

```cpp
for (const quint8 &i : ipv6.c)
    ret.append(i);
```

Also 2 final questions:

- The `// don't use auto here!` comment is purely due to coding guidelines reasons, right? Or is there any technical reason?
- Should I use `asConst(ipv6.c)`? What are the reasons for using `asConst()`? I'm only asking because I saw it in other places.

Reply:
>I can use the reference:

Please don't do that. Only use reference on types that are complex or expensive to copy.

>The // don't use auto here! comment is purely due to coding guidelines reasons, right? Or is there any technical reason?

IMO `auto` is used in place for some complex types, iterators, lambdas, etc. Here is not the case.

>Should I use asConst(ipv6.c)? What are the reasons for using asConst()? I'm only asking because I saw it in other places.

Only use it on non-const Qt containers.

Reply:
Ok, fixed. Thanks for the explanations and for bearing with me.